src/client/java/net/neoforged/neoforge/client/neo3d/GpuDeviceProperties.java

```diff
@@ -0,0 +1,156 @@
+    public final Set<SourceFactor> knownSourceFactors = defaultVal(Collections.unmodifiableSet(EnumSet.allOf(SourceFactor.class)));
+    public final Set<ShaderType> knownShaderTypes = defaultVal(Collections.unmodifiableSet(EnumSet.allOf(ShaderType.class)));
+    public final Set<UniformType> knownUniformTypes = defaultVal(Collections.unmodifiableSet(EnumSet.allOf(UniformType.class)));
+    public final Set<AddressMode> knownAddressModes = defaultVal(Collections.unmodifiableSet(EnumSet.allOf(AddressMode.class)));
+    public final Set<TextureFormat> knownTextureFormats = defaultVal(Collections.unmodifiableSet(EnumSet.allOf(TextureFormat.class)));
+
+    /**
+     * Some APIs require a 0-1 depth range, this may affect shaders that rely on specific depth ranges
+     */
+    public final boolean depthZeroToOne = defaultVal(false);
+
+    /**
+     * Total uniforms that can be used by a single program
+     * The sum of the per-type maximums may exceed this value
+     */
+    public final int maximumUniformBindings = atLeast(16);
+    /**
+     * min(GL_MAX_VERTEX_UNIFORM_BLOCKS, GL_MAX_FRAGMENT_UNIFORM_BLOCKS)
+     */
+    public final int maximumUBOBindings = atLeast(12);
+    /**
+     * Maximum number of textures that can be bound to a single pipeline.
+     * TexelBuffers also count against this limit.
+     * min(GL_MAX_VERTEX_TEXTURE_IMAGE_UNITS, GL_MAX_TEXTURE_IMAGE_UNITS)
+     */
+    public final int maximumTextureBindings = atLeast(16);
+
+    /**
+     * GL_UNIFORM_BUFFER_OFFSET_ALIGNMENT
+     * GpuDevice#getUniformOffsetAlignment()
+     * This value is known to be greater than 1 on many GPUs
+     */
+    public final int minimumUBOAlignment = minimumPossible(1);
+    /**
+     * GL_MAX_UNIFORM_BLOCK_SIZE
+     */
+    public final int maximumUBOSize = atLeast(16384);
+    /**
+     * GL_MAX_TEXTURE_BUFFER_SIZE
+     */
+    public final int maximumTexelBufferSize = atLeast(65536);
+
+    /**
+     * GL_MAX_VERTEX_ATTRIBS
+     */
+    public final int maximumVertexInputAttributes = atLeast(16);
+    /**
+     * GL_MAX_VERTEX_ATTRIB_BINDINGS
+     */
+    public final int maximumVertexInputBuffers = atLeast(16);
+    /**
+     * GL_MAX_VERTEX_OUTPUT_COMPONENTS / 4
+     */
+    public final int maximumVertexOutputLocations = atLeast(16);
+
+    /**
+     * GL_MAX_ARRAY_TEXTURE_LAYERS
+     */
+    public final int maximumImageArrayLayers = atLeast(256);
+    /**
+     * GL_MAX_TEXTURE_SIZE
+     */
+    public final int maximumImageDimension1D = atLeast(1024);
+    /**
+     * GL_MAX_TEXTURE_SIZE
+     */
+    public final int maximumImageDimension2D = atLeast(1024);
+    /**
+     * GL_MAX_3D_TEXTURE_SIZE
+     */
+    public final int maximumImageDimension3D = atLeast(1024);
+    /**
+     * GL_MAX_CUBE_MAP_TEXTURE_SIZE
+     */
+    public final int maximumImageDimensionCube = atLeast(1024);
+
+    public static <T extends GpuDeviceProperties> T create(Class<T> clazz, Map<String, Object> values) {
```

Comment:
Reflection mess again :/

Reply:
Because these are "final" they must be set via reflection, or in the constructor. Because this also needs to be binary compatible as additional fields get added, setting in constructor is not an option.

Reply:
I agree with Tech that this should be an interface. Adding new properties would work similarly to adding new feature getters to `GpuDeviceFeatures`, i.e. they get added with a default impl which returns a value that makes the property effectively absent/unsupported (i.e. zero size for a buffer or texture-like thing, etc.).